community/kernel/src/main/java/org/neo4j/kernel/impl/api/DefaultTransactionTracer.java

```diff
@@ -32,21 +33,37 @@
 import org.neo4j.kernel.impl.transaction.tracing.StoreApplyEvent;
 import org.neo4j.kernel.impl.transaction.tracing.TransactionEvent;
 import org.neo4j.kernel.impl.transaction.tracing.TransactionTracer;
+import org.neo4j.kernel.impl.util.JobScheduler;
 
-public class DefaultTransactionTracer implements TransactionTracer, LogRotationMonitor
+public class DefaultTransactionTracer implements TransactionTracer, LogRotationMonitor, LogRotationEventObservable
 {
     private final Clock clock;
+    private final JobScheduler jobScheduler;
+
     private final AtomicLong counter = new AtomicLong();
     private final AtomicLong accumulatedTotalTimeNanos = new AtomicLong();
 
+    private volatile Observer observer;
     private long startTimeNanos;
 
+    private final BiFunction<Long,Long,Runnable> notifyObserver = ( Long delta, Long count ) -> () -> {
+        Observer observer = DefaultTransactionTracer.this.observer;
+        if ( observer != null )
+        {
+            long millis = TimeUnit.NANOSECONDS.toMillis( delta );
+            observer.logRotationEvent( millis, count );
+        }
+    };
+
     private final LogRotateEvent logRotateEvent = new LogRotateEvent()
     {
         @Override
         public void close()
         {
-            accumulatedTotalTimeNanos.addAndGet( clock.nanoTime() - startTimeNanos );
+            long count = counter.incrementAndGet();
+            long delta = clock.nanoTime() - startTimeNanos;
+            accumulatedTotalTimeNanos.addAndGet( delta );
+            jobScheduler.schedule( JobScheduler.Groups.metricsEvent, notifyObserver.apply( delta, count ) );
```

Comment:
This code is on the hot-path we cannot call the metrics from this thread, hence we need to schedule one to do the job for us.